mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+	//    (Note that Spanner doesn't allow "." in table names, so this
+	//    will eventually get re-mapped when we construct the Spanner table name).
+	// d) return error if Table is nil or "".
+	var l []string
+
+	if table.Schema.String() != "" {
+		l = append(l, table.Schema.String())
+	}
+	if table.Name.String() == "" {
+		return "", fmt.Errorf("tablename is empty: can't build table name")
+	}
+	l = append(l, table.Name.String())
+	return strings.Join(l, "."), nil
+}
+
+func processColumn(conv *internal.Conv, tableName string, col *ast.ColumnDef) (string, schema.Column, bool, error) {
+	if col.Name == nil {
+		return "", schema.Column{}, false, fmt.Errorf("column name is nil")
+	}
+	name := col.Name.OrigColName()
+	if col.Tp == nil {
+		return "", schema.Column{}, false, fmt.Errorf("can't get column type for %s: %w", name, fmt.Errorf("found nil *ast.ColumnDef.Tp"))
+	}
+	tid, mods := getTypeModsAndID(col.Tp.String())
+	ty := schema.Type{
+		Name:        tid,
+		Mods:        mods,
+		ArrayBounds: getArrayBounds(col.Tp.String(), col.Tp.Elems)}
+	column := schema.Column{Name: name, Type: ty}
+	return name, column, updateColsByOption(conv, tableName, col, &column), nil
+}
+
+// updateColsByOption is specifially for ColDef constraints.
+// ColumnOption type is used for parsing column constraint info from MySQL
+func updateColsByOption(conv *internal.Conv, tableName string, col *ast.ColumnDef, column *schema.Column) bool {
+	isPk := false
+	for _, elem := range col.Options {
+		switch op := elem.Tp; op {
+		case ast.ColumnOptionPrimaryKey:
+			column.NotNull = true
+			// If primary key is defined in a column then `isPk` will be true
+			// and this column will be added in colDef as primary keys
+			isPk = true
+		case ast.ColumnOptionNotNull:
+			column.NotNull = true
+		case ast.ColumnOptionDefaultValue:
+			column.Ignored.Default = true
+		case ast.ColumnOptionUniqKey:
+			column.Unique = true
+		case ast.ColumnOptionCheck:
+			column.Ignored.Check = true
+		case ast.ColumnOptionReference:
+			column.Ignored.ForeignKey = true
+		default:
+			conv.Unexpected(fmt.Sprintf("Not supported type option of const value `%v` found in table : %s , column : %s. Mapping of columnOptionType can be found here: 'https://godoc.org/github.com/pingcap/parser/ast#ColumnOptionType'", op, tableName, column.Name))
+		}
+	}
+	return isPk
+}
+
+//getTypeModsAndID returns ID and mods of column datatype
+func getTypeModsAndID(columnType string) (string, []int64) {
+	//There are no methods in pincap parser to retirieve ID and mods
+	//We will process columnType eg:'varchar(40)' and split ID from the string.
+	//We retrieve mods using regex expression and convert it to INT64
+	id := columnType
+	var mods []int64
+	if strings.Contains(columnType, "(") {
+		id = strings.Split(columnType, "(")[0]
+		// For 'set' and 'enum' datatypes, values provided are not maxLength
+		if id == "set" || id == "enum" {
+			return id, nil
+		}
+		r, _ := regexp.Compile("\\((.*?)\\)")
```

Comment:
Use `MustCompile` and move it to a global variable, like https://github.com/cloudspannerecosystem/harbourbridge/blob/9db6997cea5ddd4ecba2775ce0c39ad380666e60/internal/fixname.go#L21. 

This will cause a runtime panic if the regular expression is invalid. If we put this into a global variable, it will be detected at compile time.

Reply:
https://github.com/cloudspannerecosystem/harbourbridge/blob/fa3d015562a86e51157a983f900b07a1fb5ca62d/mysql/mysqldump.go#L32-L33

https://github.com/cloudspannerecosystem/harbourbridge/blob/fa3d015562a86e51157a983f900b07a1fb5ca62d/mysql/mysqldump.go#L366

Done